### PR TITLE
Add more methods to Utf8String

### PR DESCRIPTION
**Pull request by krwq**

### CODE REVIEW

File: src/System.Text.Utf8/src/System/Text/Utf8/Utf8CodeUnit.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+﻿namespace System.Text.Utf8
+{
+    public struct Utf8CodeUnit : IEquatable<Utf8CodeUnit>
```
Comment (KrzysztofCwalina) [inline src/System.Text.Utf8/src/System/Text/Utf8/Utf8CodeUnit.cs]: Why do we need this to be public? Why do we need a type at all? UTF8 code unit is byte, isn't it?

File: src/System.Text.Utf8/src/System/Text/Utf8/Utf8String.cs [hunk 2]
```diff
@@ -101,6 +102,33 @@ IEnumerator IEnumerable.GetEnumerator()
...
+            {
+                return (Utf8CodeUnit)_bytes[_index + i];
+            }
+            else
+            {
+                return (Utf8CodeUnit)_buffer[i];
+            }
+        }
+
+        public Utf8CodeUnit this[int i]
```
Comment (KrzysztofCwalina) [inline src/System.Text.Utf8/src/System/Text/Utf8/Utf8String.cs]: Why do we need the indexer at all? If we need it, can we return byte?

File: src/System.Text.Utf8/src/System/Text/Utf8/Utf8CodeUnit.cs (hunk 1 above)
Comment (krwq) [inline src/System.Text.Utf8/src/System/Text/Utf8/Utf8CodeUnit.cs]: UTF-16's code unit is UInt16 and we called it char, I think it is more consistent to have some wrapper type (not pushing on the name). You could as well operate directly on byte[] and never create Utf8String and only extension methods to byte[]

Comment (KrzysztofCwalina) [inline src/System.Text.Utf8/src/System/Text/Utf8/Utf8CodeUnit.cs]: I think public types need to add enough value to justify their existence. I don't think Utf8CodeUnit adds enough value. And we are not consistent with System.String anyway. Utf8String is actually a sub-string (slice of string), it's not immutable, and amny other differences.

Comment (krwq) [inline src/System.Text.Utf8/src/System/Text/Utf8/Utf8CodeUnit.cs]: Here's the thing: Whenever we release Utf8String we will need a way of providing how to assign a single character to Utf8String. If you decide to use byte, you won't be able to implement assignment as byte is not a character, same for char, as char is not UTF8 character and thus can be multibyte. There is no standard type representing it. If you decide to use Utf8CodeUnit you can easily add compiler support for ascii characters or "surrogate" utf8 string characters. Byte seems inappropriate in this context.

Comment (krwq) [inline src/System.Text.Utf8/src/System/Text/Utf8/Utf8CodeUnit.cs]: Talked offline, we will discuss it during API Review, for now leaving as is

File: demos/LowAllocationWebServer/HttpRequestParser.cs [hunk 3]
```diff
@@ -5,49 +5,49 @@ namespace System.Net.Http.Buffered
...
-        {
-            throw new InvalidOperationException("you don't want it in a hashtable, do you?");
-        }
-
-        public override string ToString()
-        {
-            return Encoding.UTF8.GetString(_bytes.CreateArray());
-        }
-    }
+    //public struct Utf8Span
```
Comment (krwq) [inline demos/LowAllocationWebServer/HttpRequestParser.cs]: that wasn't intended in this PR. will revert Server changes

File: src/System.Text.Utf8/src/System/Text/Utf8/Utf8String.CodePointEnumerator.cs [hunk 4]
```diff
@@ -0,0 +1,164 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Comment (krwq) [inline src/System.Text.Utf8/src/System/Text/Utf8/Utf8String.CodePointEnumerator.cs]: This is the old (unchanged except for the name) Enumerator implementation. Current Enumerator is new.

**Comment by KrzysztofCwalina** [comment]
Please squash the commits and feel free to merge.

**Pull request merged**